data/src/main/java/org/apache/iceberg/data/orc/GenericOrcReader.java

```diff
@@ -172,14 +208,13 @@ public Double convert(ColumnVector vector, int row) {
     }
   }
 
-  private static class TimestampConverter implements Converter<Long> {
-    private Long convert(TimestampColumnVector vector, int row) {
-      // compute microseconds past 1970.
-      return (vector.time[row] / 1000) * 1_000_000 + vector.nanos[row] / 1000;
+  private static class TimestampTzConverter implements Converter<OffsetDateTime> {
+    private OffsetDateTime convert(TimestampColumnVector vector, int row) {
+      return EPOCH.plus(vector.time[row], ChronoUnit.MILLIS).plus(vector.nanos[row] % 1_000_000, ChronoUnit.NANOS);
```

Comment:
Seems like `vector.nanos[row]` is not absolute time in nanos and is like `java.sql.Time.getNanos`. In that case we don't need to take the `%` ?

Reply:
From https://issues.apache.org/jira/browse/ORC-546,
> Hive's TimestampColumnVector has a bad design with millis from 1970 and nanos within the second. This is consistent with java.sql.Timestamp, but it causes the millis to overlap with the nanos

Basically, the millis component of timestamp is double counted in both vector.time and also vector.nanos. Here we remove it from vector.nanos and only use it from vector.time

Reply:
I also made a small change to not use millis from vector.time and rather use millis from vector.nanos. This is consistent with other places where we are handling ORC timestamps.